### PR TITLE
SM64 - Goal Logic and Hint Bugfixes

### PR DESCRIPTION
**Pull request by Magnemania**
## What is this fixing or adding?

- Goal logic did not specify that the upper sections of Bowser in the Sky and Bowser in the Fire Sea must be reachable, creating unbeatable games on minimal accessibility.
- New subregion system broke ER hints for most of the game's location.  Using new _subregion_map_ to map those subregions to the primary level.
- Tiny-Huge Island location ER hints have been broken for several releases, added a special exception to fix.  Tiny-Huge Island locations will show both possible entrances in their hint.

## How was this tested?

- Tested goal logic by generating minimal accessibility games with move randomizer and all bowser goal and confirming that items required to reach the goals appeared in the playthrough

- Tested ER hints by generating several ER worlds, running /hint <Player> Everything, and confirming that all hints for locations within courses had entrance hints.  Collected one check in the client to confirm accuracy.

## If this makes graphical changes, please attach screenshots.

image

Pinging maintainer @N00byKing

### CODE REVIEW

**Comment by N00byKing** [comment]
Instead of an additional list, can you instead make it a property of the region? So create a SM64Region class extending Region, and add the subregion_map there. You can then set it for each (main) region. Then the list doesn't have to be passed through __init__.py and Region.py

**Comment by Silvris** [comment]
Potentially a little out of scope, but since you're adjusting the entrance hints, you might want to move it from `modify_multidata` to `extend_hint_information`. The latter runs before starting hints are generated, so currently no entrance information is placed on starting_hints.

File: worlds/sm64ex/__init__.py [hunk 1]
```diff
@@ -76,7 +77,8 @@ def generate_early(self):
         self.topology_present = self.multiworld.AreaRandomizer[self.player].value
 
     def create_regions(self):
-        create_regions(self.multiworld, self.player)
+        self.subregion_map = {}
+        create_regions(self.multiworld, self.player, self.subregion_map)
```
Comment (alwaysintreble) [inline worlds/sm64ex/__init__.py:81]: just pass the world and create/assign the dict there

**Review by alwaysintreble — COMMENTED** [review]
i've only got one comment that hasn't already been said

**Comment by Magnemania** [comment]
> Instead of an additional list, can you instead make it a property of the region? So create a SM64Region class extending Region, and add the subregion_map there. You can then set it for each (main) region. Then the list doesn't have to be passed through **init**.py and Region.py

Changed to this implementation.

> Potentially a little out of scope, but since you're adjusting the entrance hints, you might want to move it from `modify_multidata` to `extend_hint_information`. The latter runs before starting hints are generated, so currently no entrance information is placed on starting_hints.

Done.  I think that fits as part of a bugfix.

> Also probably out of scope, but since you're already changing lines with instances of this, please remove all cases of `world: MultiWorld` from this World. Also SM64 will need to be updated to the new Options API at some point.

That's a bit more out of scope, isn't it?  I believe that would involve changing every line that calls an option, so it should go with the options refactor.

File: worlds/sm64ex/__init__.py (hunk 1 above)
Comment (Magnemania) [reply worlds/sm64ex/__init__.py]: I'm going with N00byking's suggestion for implementation (as the world maintainer).

**Comment by N00byKing** [comment]
Options refactor coming soon ™️ , just need a bit more free time

**Review by N00byKing — APPROVED** [review]

**Review by RBmans — APPROVED** [review]
Everything here looks good, logic wise! Some other noncritical move rando logic changes will be coming a little later down the pipeline once some more testing/experimenting has been done, mostly to simplify logic for strict move requirements on. Or expand it, in the case of strict move requirements being turned off.

**Review by Berserker66 — APPROVED** [review]

**Pull request merged**